lib/types/rules/ecmascript-6.d.ts

```diff
@@ -291,23 +313,39 @@ export interface ECMAScript6 extends Linter.RulesRecord {
         [
             ...Array<
                 | string
-                | {
-                    name: string;
-                    importNames?: string[] | undefined;
-                    message?: string | undefined;
-                }
+                | ({
+                      name: string;
+                      message?: string;
+                      importNames?: string[];
+                      allowImportNames?: string[];
+                  } & (
+                      | { importNames?: string[]; allowImportNames?: never }
+                      | { allowImportNames?: string[]; importNames?: never }
+                  ))
```

Comment:
[Bug] `importNames` and `allowImportNames` will become `?: undefined` in this union, not `?: never`, even with [`exactOptionalPropertyTypes`](https://aka.ms/tsconfig#exactOptionalPropertyTypes) [[Playground showing `exactOptionalPropertyTypes`](https://www.typescriptlang.org/play/?exactOptionalPropertyTypes=true#code/C4TwDgpgBAKhDOwoF4oAoDeAoKUB2AhgLYQBcUiATgJZ4DmA3DlCfPAXRAPzlW2PNqRMAHtKwAHLEEPCsBr0A2gF0muAgBsNIgO4BJYWMnT4svktVYAvlABk6ZrgA+UDFCGjxU1mfn8VDFCa2vqGXiayeBAAbhCUUFaOUC5uwboGnsY+vH4WgR5G3jLkUbHxibgAlJVYWAAmEADGGgSU0I0ieIhQ0ZoArmSwCMBMWNQAZui9GgMAdAXhrJWuzNMDalAA9JtQAHpctbhrELNpoZlF8BvbuLj7hz39JwtZCNc7t-dWQA)].

This is because `never | undefined` is `undefined`, not `never`. So `importNames?: string[]` & `importNames?: never` becomes `importNames?: never`.

I'd rearrange this to something like the `ValidNoRestrictedImportPatternOptions` earlier in the file. As in, have a type for the common options (`name`, `message`) and then `&` it with other types.

Reply:
Good catch! Updated.